drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+    Irp->IoStatus.Status = status;
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+    /* TODO: protect by spinlock */
+    /* TODO: still needed? */
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+
+    /* This will create a tdiIrp: */
+    status = TdiSendDatagram(&tdiIrp, s->LocalAddressFile, ((char *)BufferData) + Buffer->Offset,
+                        Buffer->Length, TargetConnectionInfo, NetioComplete, nc);
+
+    /* TODO: check this again. I would assume if the TDI helper function
+       returns any error, the completion function is not called.
```

Comment:
The `TRUE, TRUE, TRUE` here: https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/ddk/tdikrnl.h;hb=bc76250bd8cad3883397c94ce091632e8424b066#l391
Tells you that the completion routine will get called on error.
The only cases where your completion routine won't get called is the case where `tdiIrp == NULL`, i.e. `TdiSendDatagram` fails before it even sent the IRP.

Reply:
Hi ok as I understand it the answer is yes it may fail without calling the completion routine, but only
if the allocation of the tdiIrp fails. For that case we need the manual completion of the user irp.

I changed the comment to reflect that.

Reply:
Hi I am resolving this conversation since I think I addressed what you said.